metldata/transform/base.py

```diff
@@ -97,3 +97,156 @@ class TransformationDefinition(Generic[Config]):
             "if the transformation fails."
         ),
     )
+
+
+class WorkflowConfig(BaseModel, ABC):
+    """A base class for workflow configs."""
+
+
+class WorkflowStepBase(BaseModel, ABC):
+    """A base class for workflow steps."""
+
+    description: str = Field(..., description="A description of the step.")
+    input: Optional[str] = Field(
+        ...,
+        description=(
+            "The name of the workflow step from which the output is used as input"
+            " for this step. If this is the first step, set to None."
+        ),
+    )
+
+    class Config:
+        """Config for the workflow step."""
+
+        frozen = True
+
+
+class WorkflowStep(WorkflowStepBase):
+    """A single step in a transformation workflow."""
+
+    transformation_definition: TransformationDefinition = Field(
+        ...,
+        description="The transformation to be executed in this step.",
+    )
+
+
+class WorkflowDefinition(BaseModel):
+    """A definition of a transformation workflow."""
+
+    description: str = Field(..., description="A description of the workflow.")
+    steps: dict[str, WorkflowStep] = Field(
+        ...,
+        description=(
+            "A dictionary of workflow steps. The keys are the names of the steps, and"
+            + " the values are the workflow steps themselves."
+        ),
+    )
+    artifacts: dict[str, str] = Field(
+        ...,
+        description=(
+            "A dictionary of artifacts that are output by this workflow."
+            + " The keys are the names of the artifacts, and the values are the names"
+            + " of the workflow steps that output them."
+        ),
+    )
+
+    # pylint: disable=no-self-argument
+    @validator("steps", pre=False)
+    def validate_step_references(
+        cls, steps: dict[str, WorkflowStep]
+    ) -> dict[str, WorkflowStep]:
+        """Validate that workflow steps reference other existing other steps as input.
+        There should be exactly one step with input=None.
+        """
+
+        step_with_no_input_found = False
+
+        for step_name, step in steps.copy().items():
```

Comment:
```suggestion
        for step_name, step in steps.items():
```

Copy should not be needed since nothing is modified.